geopandas/tools/clip.py

```diff
@@ -0,0 +1,183 @@
+    clip_obj : GeoDataFrame, GeoSeries, (Multi)Polygon
+          Polygon vector layer used to clip gdf.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with gdf.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from gdf clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clip points (global capital cities) with a polygon (the South American continent):
+
+        >>> import geopandas as gpd
+        >>> import geopandas.clip as gc
+        >>> path = geopandas.datasets.get_path('naturalearth_lowres')
+        >>> world = geopandas.read_file(path)
+        >>> south_america = world[world['continent'] == "South America"]
+        >>> capitals = gpd.read_file(gpd.datasets.get_path('naturalearth_cities'))
+        >>> capitals.shape
+        (202, 2)
+        >>> sa_capitals = cl.clip(capitals, south_america)
+        >>> sa_capitals.shape
+        (12, 2)
+    """
+    if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
+
+    if not isinstance(clip_obj, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon)):
+        raise TypeError(
+            "'clip_obj' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(gdf))
+        )
+
+    if isinstance(clip_obj, (GeoDataFrame, GeoSeries)):
+        xmin, ymin, xmax, ymax = clip_obj.total_bounds
+    else:
+        xmin, ymin, xmax, ymax = clip_obj.bounds
+    if gdf.cx[xmin:xmax, ymin:ymax].empty:
+        raise ValueError("Shape and crop extent do not overlap.")
+
+    if isinstance(clip_obj, (GeoDataFrame, GeoSeries)):
+        poly = clip_obj.geometry.unary_union
+    else:
+        poly = clip_obj
+
+    geom_types = gdf.geometry.type
+    poly_line_idx = np.asarray(
+        (geom_types == "LineString")
+        | (geom_types == "LinearRing")
+        | (geom_types == "Polygon")
+        | (geom_types == "MultiPolygon")
+        | (geom_types == "MultiLineString")
+    )
+
+    point_idx = np.asarray((geom_types == "Point") | (geom_types == "MultiPoint"))
+
+    points = gdf[point_idx]
+    if not points.empty:
+        point_gdf = _clip_points(points, poly)
+    else:
+        point_gdf = None
+
+    poly_lines = gdf[poly_line_idx]
+    if not poly_lines.empty:
+        poly_line_gdf = _clip_line_poly(poly_lines, poly)
+    else:
+        poly_line_gdf = None
+
+    order = pd.Series(range(len(gdf)), index=gdf.index)
+    concat = pd.concat([point_gdf, poly_line_gdf])
+    concat["_order"] = order
+
+    if "GeometryCollection" in concat.geom_type[0]:
+        warnings.warn(
+            "A geometry collection has been returned. This is likely due to a"
+            "line sliver being returned after clipping a polygon. To properly "
+            "plot your data you can use .explode()"
```

Comment:
I am gonna fix GeometryCollection plotting bug for 0.7.0. In fact, I already did I just did not have time to write tests for it. I would change this warning to a more general one. Something like `To avoid GeometryCollection you can use .explode().`

Reply:
@martinfleis we are adding the drop_slivers functionality now. so it seems like we should mention that here. how about something like:

```
A geometry collection has been returned. Use .explode() to remove the collection object or drop_slivers=True to remove sliver geometries."
```